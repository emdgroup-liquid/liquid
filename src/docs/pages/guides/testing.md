---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

