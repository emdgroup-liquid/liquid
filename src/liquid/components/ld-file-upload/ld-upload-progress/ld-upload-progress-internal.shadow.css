:host {
  display: flex;
}

.ld-upload-progress {
  display: flex;
  list-style: none;
  margin: 0;
  padding: var(--ld-upload-progress-padding, 0);
  width: var(--ld-upload-progress-width, 100%);
  max-height: var(--ld-upload-progress-max-height, 22rem);
  overflow: hidden;
  overflow-y: scroll;
  border: var(--ld-sp-2) solid var(--ld-col-neutral-100);
  border-radius: var(--ld-upload-progress-border-radius, var(--ld-br-l));
  gap: var(--ld-sp-2);
  background-color: var(
    --ld-upload-progress-background-color,
    var(--ld-col-neutral-100)
  );
  flex-direction: column;
}
