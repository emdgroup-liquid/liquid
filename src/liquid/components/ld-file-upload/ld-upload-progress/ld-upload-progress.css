:host {
  display: flex;
}

.ld-upload-progress {
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0;
  width: 100%;
  max-height: 22rem;
  overflow: hidden;
  overflow-y: scroll;
  border: var(--ld-sp-2) solid var(--ld-col-neutral-100);
  border-radius: var(--ld-br-l);
  gap: var(--ld-sp-2);
  background-color: var(--ld-col-neutral-100);
  flex-direction: column;
}
