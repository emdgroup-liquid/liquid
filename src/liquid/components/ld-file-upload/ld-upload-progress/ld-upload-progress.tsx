import { Component, Element, h, Host, Prop } from '@stencil/core'
import type { UploadItem } from '../ld-file-upload'
import { LdUploadItemConfig } from '../ld-upload-item/ld-upload-item.types'

/**
 * @virtualProp ref - reference to component
 */
@Component({
  tag: 'ld-upload-progress',
  styleUrl: 'ld-upload-progress.css',
  shadow: true,
})
export class LdUploadProgress {
  @Element() el: HTMLLdUploadProgressElement

  /** Size of the context menu. */
  @Prop() size?: 'sm' | 'lg'

  /** startUpload defines whether upload starts immediately after choosing files or after confirmation. */
  @Prop() startUpload?: boolean = false

  /** allowPause defines whether the user will be able to pause uploads. */
  @Prop() allowPause?: boolean

  /** showTotalProgress defines whether the total progress of all upoading files will be shown in the progress button */
  @Prop() showProgress?: boolean = false

  /** Maps file types to icon path */
  @Prop() icons?: Partial<LdUploadItemConfig>

  /** List of files */
  @Prop() uploadItems: UploadItem[] = []

  /* @Prop() uploadItemTypes: {
    fileName: string
    fileType: string
  }[] = [] */

  private renderListItems = () =>
    this.uploadItems.map((item) => (
      <li>
        <ld-upload-item
          key={item.fileName + '_' + item.fileSize + '_' + item.progress}
          state={item.state}
          fileName={item.fileName}
          fileSize={item.fileSize}
          fileType={item.fileType}
          progress={item.progress}
          allowPause={this.allowPause}
          showProgress={this.showProgress}
          icons={this.icons}
        ></ld-upload-item>
      </li>
    ))

  render() {
    return (
      <Host>
        <ul class="ld-upload-progress">{this.renderListItems()}</ul>
      </Host>
    )
  }
}
