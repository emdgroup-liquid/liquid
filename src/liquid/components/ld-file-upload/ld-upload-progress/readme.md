---
eleventyNavigation:
  key: Upload Progress
  parent: File Upload
layout: layout.njk
title: Upload Progress
permalink: components/ld-file-upload/ld-upload-progress/
---

# ld-upload-progress

The `ld-upload-progress` component is used internally for the `ld-file-upload`. It is the visual part of the list of files and their current upload progress.

## Examples

### Default

{% example '{ "opened": true }' %}
<ld-upload-progress start-upload='false'>
<ld-upload-item state='pending' file-name='Liquid' file-size='1.28'></ld-upload-item>
<ld-upload-item file-name='Liquid' file-size='1.28'></ld-upload-item>
<ld-upload-item file-name='Liquid' file-size='1.28'></ld-upload-item>
<ld-upload-item file-name='Liquid' file-size='1.28'></ld-upload-item>
</ld-upload-progress>

<!-- React component -->

<!-- CSS component -->

{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                                                                                                     | Type                                                                                       | Default     |
| -------------- | --------------- | --------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------ | ----------- |
| `allowPause`   | `allow-pause`   | allowPause defines whether the user will be able to pause uploads.                                              | `boolean`                                                                                  | `undefined` |
| `icons`        | --              | Maps file types to icon path                                                                                    | `{ pdf?: string; zip?: string; jpeg?: string; txt?: string; png?: string; rtf?: string; }` | `undefined` |
| `ref`          | `ref`           | reference to component                                                                                          | `any`                                                                                      | `undefined` |
| `showProgress` | `show-progress` | showTotalProgress defines whether the total progress of all upoading files will be shown in the progress button | `boolean`                                                                                  | `false`     |
| `size`         | `size`          | Size of the context menu.                                                                                       | `"lg" \| "sm"`                                                                             | `undefined` |
| `startUpload`  | `start-upload`  | startUpload defines whether upload starts immediately after choosing files or after confirmation.               | `boolean`                                                                                  | `false`     |
| `uploadItems`  | --              | List of files                                                                                                   | `UploadItem[]`                                                                             | `[]`        |


## Dependencies

### Used by

 - [ld-file-upload](..)

### Depends on

- [ld-upload-item](../ld-upload-item)

### Graph
```mermaid
graph TD;
  ld-upload-progress --> ld-upload-item
  ld-upload-item --> ld-icon
  ld-upload-item --> ld-typo
  ld-upload-item --> ld-tooltip
  ld-upload-item --> ld-button
  ld-upload-item --> ld-sr-only
  ld-upload-item --> ld-progress
  ld-upload-item --> ld-loading
  ld-upload-item --> ld-input-message
  ld-tooltip --> ld-sr-only
  ld-tooltip --> ld-tooltip-popper
  ld-input-message --> ld-icon
  ld-file-upload --> ld-upload-progress
  style ld-upload-progress fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
