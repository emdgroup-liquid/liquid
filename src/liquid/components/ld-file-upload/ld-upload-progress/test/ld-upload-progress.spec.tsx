import { newSpecPage } from '@stencil/core/testing'
import { LdUploadProgress } from '../ld-upload-progress'

describe('ld-upload-progress', () => {
  it('renders default', async () => {
    const page = await newSpecPage({
      components: [LdUploadProgress],
      html: `<ld-upload-progress></ld-upload-progress>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})

describe('ld-upload-progress', () => {
  it('renders list item', async () => {
    const page = await newSpecPage({
      components: [LdUploadProgress],
      html: `<ld-upload-progress uploadItems=[
        {
          state: 'uploading',
          fileName: 'file1.png',
          fileSize: 100000,
          fileType: 'png',
          progress: 50,
        },]>
        </ld-upload-progress>`,
    })
    expect(page.root).toMatchSnapshot()
  })
})
