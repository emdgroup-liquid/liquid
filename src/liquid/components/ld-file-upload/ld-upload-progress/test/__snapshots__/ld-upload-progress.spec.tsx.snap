// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-upload-progress renders default 1`] = `
<ld-upload-progress>
  <mock:shadow-root>
    <ul class="ld-upload-progress"></ul>
  </mock:shadow-root>
</ld-upload-progress>
`;

exports[`ld-upload-progress renders list item 1`] = `
<ld-upload-progress 'file1.png',="" 'png',="" 'uploading',="" 100000,="" 50,="" filename:="" filesize:="" filetype:="" progress:="" state:="" uploaditems="[" {="" },]="">
  <mock:shadow-root>
    <ul class="ld-upload-progress"></ul>
  </mock:shadow-root>
</ld-upload-progress>
`;
